### ISSUE TITLE
Gradient cannot deal with cyclic dependencies (recursive types)

### ISSUE THREAD

**Issue by eksperimental** [esl/gradient#129]
I managed to track down what was the issue causing what I have previously reported in https://github.com/esl/gradient/issues/37

Gradient (and or Gradualizer) cannot deal with cyclic dependencies  (recursive types), for example:
```elixir
@spec list_flatten(deep_list) :: list when deep_list: [any | deep_list]
```

Here's an example that isolates the issue.
The code is the copy of `List.flatten/1,2` in Elixir which delegates to `:lists.flatten/1,2` in Erlang.

The typespecs of the Elixir functions are analogous to the Erlang ones (I have manually translated them in order to verify this and they are the same). 

https://github.com/elixir-lang/elixir/blob/aabfd191f983f51610b554d596887a32b7d30c69/lib/elixir/lib/list.ex#L247
https://github.com/elixir-lang/elixir/blob/aabfd191f983f51610b554d596887a32b7d30c69/lib/elixir/lib/list.ex#L265

Here's the Elixir code which could be considered a minim case:

```elixir
defmodule GradientDebug do
  @spec list_flatten(deep_list) :: list when deep_list: [any | deep_list]
  def list_flatten(list) do
    :lists.flatten(list)
  end

  @spec list_flatten(deep_list, [elem]) :: [elem] when elem: var, deep_list: [elem | deep_list]
  def list_flatten(list, tail) do
    :lists.flatten(list, tail)
  end
end
```

and here's the output of Gradient
```
lib/gradient_debug.ex: The type spec bounded_fun((deep_list -> list()), [
  {:type, 0, :constraint,
   [
     {:atom, 0, :is_subtype},
     [
       {:var, 0, :deep_list},
       {:type, 0, :list, [{:type, 0, :union, [{:type, 0, :any, []}, {:var, 0, :deep_list}]}]}
     ]
   ]}
]) has a cyclic dependency in variable deep_list

lib/gradient_debug.ex: The type spec bounded_fun((deep_list, list(elem) -> list(elem)), [
  {:type, 0, :constraint,
   [
     {:atom, 0, :is_subtype},
     [
       {:var, 0, :deep_list},
       {:type, 0, :list, [{:type, 0, :union, [{:var, 0, :elem}, {:var, 0, :deep_list}]}]}
     ]
   ]}
]) has a cyclic dependency in variable deep_list

Total errors: 2
```

So basically, when you call (not when a function returns) `List.flatten/1,2` you will get this warning in Dialyzer.

**Comment by eksperimental** [comment]
Sorry, i accidentally hit submit before I finished writing the issue. Please read it again is the text was not complete https://github.com/esl/gradient/issues/129#issue-1381071516

**Comment by erszcz** [comment]
Hey, @eksperimental!

Thanks for the report. Yes, we're aware of upstream issues with handling recursive types and we're [making](https://github.com/josefs/Gradualizer/pull/368) [some](https://github.com/josefs/Gradualizer/pull/383) [progress](https://github.com/josefs/Gradualizer/pull/452) on it. This is tricky, though, as the Erlang/Elixir type language allows for both union types and recursive types, which makes expanding unions to simplify them (for example to merge integer ranges) a delicate matter. Step by step, we're going to solve it 👍

**Comment by erszcz** [comment]
Hi, @eksperimental!

With #130 merged all the examples you provide type check. However, I modified them a little bit and I also inlined the examples from `Elixir.List` you link to:

```elixir
defmodule GradientDebug do
  @type deep_list(a) :: [a | deep_list(a)]

  @spec list_flatten(deep_list(any)) :: list()
  def list_flatten(list) do
    :lists.flatten(list)
  end

  @spec list_flatten(deep_list(any), list()) :: list()
  def list_flatten(list, tail) do
    :lists.flatten(list, tail)
  end

  @spec foldl([elem], acc, (elem, acc -> acc)) :: acc when elem: var, acc: var
  def foldl(list, acc, fun) when is_list(list) and is_function(fun) do
    :lists.foldl(fun, acc, list)
  end

  @spec foldr([elem], acc, (elem, acc -> acc)) :: acc when elem: var, acc: var
  def foldr(list, acc, fun) when is_list(list) and is_function(fun) do
    :lists.foldr(fun, acc, list)
  end
end
```

The reason for this modification is that, strictly speaking:

```
@spec list_flatten(deep_list) :: list when deep_list: [any | deep_list]
```

does not use just recursive types, but recursive bounds. The difference stems from the fact that `when` does not mean just a type alias - it's a restriction on the polymorphic type variable. The wording in https://hexdocs.pm/elixir/1.13/typespecs.html is a bit less specific than in https://www.erlang.org/doc/reference_manual/typespec.html, but the latter says that `::` means «is a subtype of», that is it's an upper bound of a polymorphic type variable.

With all that in mind - recursive bounds are not supported by Gradient and Gradualizer. However, the `deep_list(a)` type we define above is sufficient to express a deeply nested list. After the modification the above snippet type checks with no errors:

```
$ ./gradient debug.ex
Typechecking files...
No errors found!
```

Thanks again for the report!
